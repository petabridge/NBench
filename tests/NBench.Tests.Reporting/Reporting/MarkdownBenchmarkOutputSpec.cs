using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using NBench.Collection;
using NBench.Metrics;
using NBench.Metrics.Counters;
using NBench.Metrics.GarbageCollection;
using NBench.Metrics.Memory;
using NBench.Reporting;
using NBench.Reporting.Targets;
#if !CORECLR
using ApprovalTests;
using ApprovalTests.Reporters;
using ApprovalTests.Scrubber;
#endif
using NBench.Sdk;
using NBench.Sys;
using Xunit;
using Xunit.Abstractions;

namespace NBench.Tests.End2End.Reporting
{
#if !CORECLR
    [UseReporter(typeof(DiffReporter))]
    public class MarkdownBenchmarkOutputSpec
    {
        private readonly ITestOutputHelper _output;
        private static readonly string _solutionPath =
            ".." + Path.DirectorySeparatorChar
            + ".." + Path.DirectorySeparatorChar
            + ".." + Path.DirectorySeparatorChar
            + ".." + Path.DirectorySeparatorChar;
        private static readonly string _perfResultsPath =
            Path.GetFullPath(_solutionPath + "PerfResults");

        public MarkdownBenchmarkOutputSpec(ITestOutputHelper ouput)
        {
            _output = ouput;
        }

        [Fact]
        public void Should_report_valid_markdown_format()
        {
            CleanPerfDir(_perfResultsPath);
            SystemTime.UtcNow = () => new DateTime(2017, 3, 13);

            var mdOutput = new MarkdownBenchmarkOutput(_perfResultsPath);
            var fakeBenchmarkResults = new BenchmarkResults("NBench.FakeBenchmark",
                new BenchmarkSettings(TestMode.Test, RunMode.Iterations, 30, 1000,
                    new List<IBenchmarkSetting>(),
                    new Dictionary<MetricName, MetricsCollectorSelector>()),
                new List<BenchmarkRunReport>()
                {
                    new BenchmarkRunReport(TimeSpan.FromSeconds(3),
                        new List<MetricRunReport>()
                        {
                            new MetricRunReport(new CounterMetricName("FakeCounterMetric"), "bytes", 0d, Stopwatch.Frequency),
                            new MetricRunReport(new GcMetricName(GcMetric.TotalCollections, GcGeneration.Gen2), "collections", 0d, Stopwatch.Frequency),
                            new MetricRunReport(new MemoryMetricName(MemoryMetric.TotalBytesAllocated), "operations", 0d, Stopwatch.Frequency),
                        },
                        new List<Exception>())
                });
            var fakeBenchmarkFinalResults = new BenchmarkFinalResults(fakeBenchmarkResults, new List<AssertionResult>()
            {
                new AssertionResult(new CounterMetricName("FakeCounterMetric"), "[Counter] FakeCounterMetric Assertion Result", true),
                new AssertionResult(new GcMetricName(GcMetric.TotalCollections, GcGeneration.Gen2), "TotalCollections [Gen2] Assertion Result", true),
                new AssertionResult(new MemoryMetricName(MemoryMetric.TotalBytesAllocated), "TotalBytesAllocated Assertion Result", true)
            });
            mdOutput.WriteBenchmark(fakeBenchmarkFinalResults);

            var fakePerfResultsFile = Directory.GetFiles(_perfResultsPath, "NBench.FakeBenchmark*",
                SearchOption.AllDirectories);

            Approvals.Verify(File.ReadAllText(fakePerfResultsFile.FirstOrDefault()), ScrubSysInfo);
        }

        private static void CleanPerfDir(string path)
        {
            if (Directory.Exists(path)) // no need to clean the directory if we don't have it
            {
                var files = Directory.GetFiles(path, "NBench.FakeBenchmark*", SearchOption.AllDirectories);
                foreach (var file in files)
                {
                    File.Delete(file);
                }
            }
        }

        private static string ScrubSysInfo(string input)
        {
            var startIndex = input.IndexOf(@"```ini");
            var endIndex = input.IndexOf(@"```" + Environment.NewLine, startIndex) + 3;
            return input.Remove(startIndex, endIndex - startIndex);
        }
    }
#endif
}
