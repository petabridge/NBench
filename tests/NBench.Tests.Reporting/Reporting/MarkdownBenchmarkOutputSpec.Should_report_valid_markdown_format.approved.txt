# NBench.FakeBenchmark
_3/13/2017 12:00:00 AM_
### System Info


### NBench Settings
```ini
RunMode=Iterations, TestMode=Test
NumberOfIterations=30, MaximumRunTime=00:00:01
Concurrent=False
Tracing=False
```

## Data
-------------------

### Totals
|          Metric |           Units |             Max |         Average |             Min |          StdDev |
|---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
|[Counter] FakeCounterMetric |           bytes |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalCollections [Gen2] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalBytesAllocated |      operations |            0.00 |            0.00 |            0.00 |            0.00 |

### Per-second Totals
|          Metric |       Units / s |         Max / s |     Average / s |         Min / s |      StdDev / s |
|---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
|[Counter] FakeCounterMetric |           bytes |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalCollections [Gen2] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalBytesAllocated |      operations |            0.00 |            0.00 |            0.00 |            0.00 |

### Raw Data
#### [Counter] FakeCounterMetric
|           Run # |           bytes |       bytes / s |      ns / bytes |
|---------------- |---------------- |---------------- |---------------- |
|               1 |            0.00 |            0.00 |1,000,000,000.00 |

#### TotalCollections [Gen2]
|           Run # |     collections | collections / s |ns / collections |
|---------------- |---------------- |---------------- |---------------- |
|               1 |            0.00 |            0.00 |1,000,000,000.00 |

#### TotalBytesAllocated
|           Run # |      operations |  operations / s | ns / operations |
|---------------- |---------------- |---------------- |---------------- |
|               1 |            0.00 |            0.00 |1,000,000,000.00 |


## Benchmark Assertions

* [Counter] FakeCounterMetric Assertion Result
* TotalCollections [Gen2] Assertion Result
* TotalBytesAllocated Assertion Result

